File: icore/src/main/java/com/ddu/icore/app/BaseApp.java
package com.ddu.icore.app;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class BaseApp extends Application implements Application.ActivityLifecycleCallbacks {

    @NonNull
    protected static Handler mainThreadHandler = new Handler();

    protected static BaseApp mApplication;

    protected WeakReference<Activity> sCurrentActivity;

    protected Map<Integer, WeakReference<Activity>> sCacheActivities;

    protected static Context mContext;

    @Override
    public void onCreate() {
        super.onCreate();
        mApplication = this;
        mContext = getApplicationContext();
    }

    public static BaseApp getApp() {
        return mApplication;
    }

    public static Context getContext() {
        return mContext;
    }

    @NonNull
    public static Handler getMainThreadHandler() {
        return mainThreadHandler;
    }

    public static void post(Runnable r) {
        mainThreadHandler.post(r);
    }

    public static void postDelayed(Runnable r, long delayMillis) {
        mainThreadHandler.postDelayed(r, delayMillis);
    }

    public Map<Integer, WeakReference<Activity>> getCacheActivities() {
        return sCacheActivities;
    }

    public WeakReference<Activity> getCurrentActivity() {
        return sCurrentActivity;
    }

    public void addActivity(@NonNull Activity activity) {
        sCurrentActivity = new WeakReference<>(activity);
        if (sCacheActivities != null) {
            int hashCode = activity.hashCode();
            if (sCacheActivities.containsKey(hashCode)) {
                sCacheActivities.remove(hashCode);
            }
            sCacheActivities.put(hashCode, new WeakReference<>(activity));
        }
    }

    public void removeActivity(@NonNull Activity activity) {
        int hashCode = activity.hashCode();
        if (sCacheActivities != null && sCacheActivities.containsKey(hashCode)) {
            sCacheActivities.remove(hashCode);
        }
    }

    @Nullable
    public Activity getCacheActivity(int hashCode) {
        WeakReference<Activity> weakReference = sCacheActivities.get(hashCode);
        if (weakReference == null) {
            return null;
        }
        return weakReference.get();
    }

    public int finishAllActivity() {
        int finishCount = 0;
        if (sCacheActivities != null && !sCacheActivities.isEmpty()) {
            List<WeakReference<Activity>> activities = new ArrayList<WeakReference<Activity>>(sCacheActivities.values());
            for (WeakReference<Activity> activity : activities) {
                Activity tempActivity = activity.get();
                if (tempActivity == null) {
                    continue;
                }
                sCacheActivities.remove(tempActivity.hashCode());
                if (tempActivity != sCurrentActivity.get()) {
                    if (tempActivity != null && !tempActivity.isFinishing()) {
                        tempActivity.finish();
                        finishCount++;
                    }
                }
            }
        }
        return finishCount;
    }

    @Override
    public void onActivityCreated(@NonNull Activity activity, Bundle savedInstanceState) {
        addActivity(activity);
    }

    @Override
    public void onActivityStarted(Activity activity) {

    }

    @Override
    public void onActivityResumed(Activity activity) {

    }

    @Override
    public void onActivityPaused(Activity activity) {

    }

    @Override
    public void onActivityStopped(Activity activity) {

    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

    }

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {
        removeActivity(activity);
    }
}

File: icore/src/main/java/com/ddu/icore/app/BaseApp.kt
package com.ddu.icore.app

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import java.lang.ref.WeakReference
import java.util.*

open class BaseApp : Application(), Application.ActivityLifecycleCallbacks {

    lateinit var currentActivity: WeakReference<Activity>

    val sCacheActivities: MutableMap<Int, WeakReference<Activity>> by lazy {
        mutableMapOf<Int, WeakReference<Activity>>()
    }

    val cacheActivities: Map<Int, WeakReference<Activity>>?
        get() = sCacheActivities

    override fun onCreate() {
        super.onCreate()
        mApp = this
    }

    open fun addActivity(activity: Activity) {
        currentActivity = WeakReference(activity)
        val hashCode = activity.hashCode()
        if (sCacheActivities.containsKey(hashCode)) {
            sCacheActivities.remove(hashCode)
        }
        sCacheActivities.put(hashCode, WeakReference(activity))
    }

    open fun removeActivity(activity: Activity) {
        val hashCode = activity.hashCode()
        if (sCacheActivities.containsKey(hashCode)) {
            sCacheActivities.remove(hashCode)
        }
    }

    open fun getCacheActivity(hashCode: Int): Activity? {
        val weakReference = sCacheActivities[hashCode] ?: return null
        return weakReference.get()
    }

    open fun finishAllActivity(): Int {
        var finishCount = 0
        if (sCacheActivities != null && !sCacheActivities.isEmpty()) {
            val activities = ArrayList(sCacheActivities.values)
            for (activity in activities) {
                val tempActivity = activity.get() ?: continue
                sCacheActivities.remove(tempActivity.hashCode())
                if (tempActivity !== currentActivity.get()) {
                    if (tempActivity != null && !tempActivity.isFinishing) {
                        tempActivity.finish()
                        finishCount++
                    }
                }
            }
        }
        return finishCount
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        addActivity(activity)
    }

    override fun onActivityStarted(activity: Activity) {

    }

    override fun onActivityResumed(activity: Activity) {

    }

    override fun onActivityPaused(activity: Activity) {

    }

    override fun onActivityStopped(activity: Activity) {

    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {

    }

    override fun onActivityDestroyed(activity: Activity) {
        removeActivity(activity)
    }

    companion object {

        lateinit var mApp: Application

        private var mainHandler = Handler(Looper.getMainLooper())

        fun post(r: Runnable) {
            mainHandler.post(r)
        }

        fun postDelayed(r: Runnable, delayMillis: Long) {
            mainHandler.postDelayed(r, delayMillis)
        }

        fun getContext(): Application {
            return mApp
        }
    }
}
